From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrii Sultanov <andriy.sultanov@vates.tech>
Date: Tue, 11 Aug 2026 09:17:14 +0000
Subject: [PATCH] ocaml/libs: Move Vhd_qcow_parsing into a library outside of
 xapi

It will be used in sparse_dd, so move it to libs/vhd

Signed-off-by: Andrii Sultanov <andriy.sultanov@vates.tech>
---
 dune-project                                  | 23 +++++++++
 ocaml/libs/vhd/vhd_qcow_parsing/dune          | 13 +++++
 .../vhd/vhd_qcow_parsing}/vhd_qcow_parsing.ml | 42 +++++++++++++++++
 .../vhd_qcow_parsing}/vhd_qcow_parsing.mli    |  9 +++-
 ocaml/xapi/dune                               |  1 +
 ocaml/xapi/qcow_tool_wrapper.ml               | 47 +++----------------
 ocaml/xapi/qcow_tool_wrapper.mli              |  4 --
 ocaml/xapi/stream_vdi.ml                      |  4 +-
 opam/vhd-tool.opam                            |  1 +
 opam/vhd_qcow_parsing.opam                    | 36 ++++++++++++++
 opam/xapi.opam                                |  1 +
 11 files changed, 132 insertions(+), 49 deletions(-)
 create mode 100644 ocaml/libs/vhd/vhd_qcow_parsing/dune
 rename ocaml/{xapi => libs/vhd/vhd_qcow_parsing}/vhd_qcow_parsing.ml (75%)
 rename ocaml/{xapi => libs/vhd/vhd_qcow_parsing}/vhd_qcow_parsing.mli (81%)
 create mode 100644 opam/vhd_qcow_parsing.opam

diff --git a/dune-project b/dune-project
index 09fa04faa..9d14bc44b 100644
--- a/dune-project
+++ b/dune-project
@@ -371,6 +371,25 @@
   (xapi-test-utils :with-test)
   (xenstore_transport :with-test)))
 
+(package
+ (name vhd_qcow_parsing)
+ (synopsis "Wrapper to parse VHD and QCOW headers to determine allocated
+           clusters")
+ (depends
+  dune
+  (ocaml (>= "4.13.0"))
+  (forkexec
+   (= :version))
+  (xapi-consts
+   (= :version))
+  (xapi-log
+   (= :version))
+  (xapi-stdext-pervasives
+   (= :version))
+  yojson
+ )
+)
+
 (package
  (name xapi)
  (synopsis "The toolstack daemon which implements the XenAPI")
@@ -490,6 +509,8 @@
   xmlm
   (xml-light2
    (= :version))
+  (vhd_qcow_parsing
+   (= :version))
   yojson
   (zstd
    (= :version))))
@@ -531,6 +552,8 @@
    (= :version))
   (vhd-format-lwt
    (= :version))
+  (vhd_qcow_parsing
+   (= :version))
   (xapi-idl
    (= :version))
   (xapi-log
diff --git a/ocaml/libs/vhd/vhd_qcow_parsing/dune b/ocaml/libs/vhd/vhd_qcow_parsing/dune
new file mode 100644
index 000000000..c8410ec8a
--- /dev/null
+++ b/ocaml/libs/vhd/vhd_qcow_parsing/dune
@@ -0,0 +1,13 @@
+(library
+  (name vhd_qcow_parsing)
+  (public_name vhd_qcow_parsing)
+  (modules vhd_qcow_parsing)
+  (libraries
+    threads.posix
+    xapi-log
+    forkexec
+    yojson
+    xapi-consts
+    xapi-stdext-pervasives
+  )
+)
diff --git a/ocaml/xapi/vhd_qcow_parsing.ml b/ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.ml
similarity index 75%
rename from ocaml/xapi/vhd_qcow_parsing.ml
rename to ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.ml
index 90ed8a7b2..86540cc1b 100644
--- a/ocaml/xapi/vhd_qcow_parsing.ml
+++ b/ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.ml
@@ -105,3 +105,45 @@ let parse_header_interval pipe_reader =
     )
   in
   (cluster_size, cluster_list)
+
+module Qcow = struct
+  let qemu_img = "/usr/lib64/xen/bin/qemu-img"
+
+  let read_header qcow_path =
+    let progress_cb _ = () in
+    let run_in_thread tool args pipe_writer replace_fds =
+      Thread.create
+        (fun () ->
+          Xapi_stdext_pervasives.Pervasiveext.finally
+            (fun () ->
+              run_tool tool progress_cb args ~output_fd:pipe_writer ~replace_fds
+            )
+            (fun () -> Unix.close pipe_writer)
+        )
+        ()
+    in
+
+    let map_pipe_reader, map_pipe_writer = Unix.pipe ~cloexec:true () in
+    let (_ : Thread.t) =
+      run_in_thread qemu_img
+        ["map"; qcow_path; "--output=json"]
+        map_pipe_writer []
+    in
+
+    let info_pipe_reader, info_pipe_writer = Unix.pipe ~cloexec:true () in
+    let (_ : Thread.t) =
+      run_in_thread qemu_img
+        ["info"; qcow_path; "--output=json"]
+        info_pipe_writer []
+    in
+
+    (map_pipe_reader, info_pipe_reader)
+
+  let parse_header qcow_path =
+    let pipe, _ = read_header qcow_path in
+    parse_header pipe
+
+  let parse_header_interval qcow_path =
+    let pipes = read_header qcow_path in
+    parse_header_qemu_img pipes
+end
diff --git a/ocaml/xapi/vhd_qcow_parsing.mli b/ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.mli
similarity index 81%
rename from ocaml/xapi/vhd_qcow_parsing.mli
rename to ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.mli
index f43b56dab..b990ea855 100644
--- a/ocaml/xapi/vhd_qcow_parsing.mli
+++ b/ocaml/libs/vhd/vhd_qcow_parsing/vhd_qcow_parsing.mli
@@ -25,5 +25,10 @@ val parse_header : Unix.file_descr -> int * int list
 
 val parse_header_interval : Unix.file_descr -> int * (int * int) list
 
-val parse_header_qemu_img :
-  Unix.file_descr * Unix.file_descr -> int * (int * int) list
+module Qcow : sig
+  val read_header : string -> Unix.file_descr * Unix.file_descr
+
+  val parse_header : string -> int * int list
+
+  val parse_header_interval : string -> int * (int * int) list
+end
diff --git a/ocaml/xapi/dune b/ocaml/xapi/dune
index b2c78a30c..961d8eed1 100644
--- a/ocaml/xapi/dune
+++ b/ocaml/xapi/dune
@@ -195,6 +195,7 @@
   uri
   uuid
   uuidm
+  vhd_qcow_parsing
   x509
   xapi_aux
   xapi-backtrace
diff --git a/ocaml/xapi/qcow_tool_wrapper.ml b/ocaml/xapi/qcow_tool_wrapper.ml
index 652db754a..d6f05aa2a 100644
--- a/ocaml/xapi/qcow_tool_wrapper.ml
+++ b/ocaml/xapi/qcow_tool_wrapper.ml
@@ -21,45 +21,6 @@ let receive (progress_cb : int -> unit) (unix_fd : Unix.file_descr)
   let qcow_tool = !Xapi_globs.qcow_stream_tool in
   Vhd_qcow_parsing.run_tool qcow_tool progress_cb args ~input_fd:unix_fd
 
-let read_header qcow_path =
-  let progress_cb _ = () in
-  let run_in_thread tool args pipe_writer replace_fds =
-    Thread.create
-      (fun () ->
-        Xapi_stdext_pervasives.Pervasiveext.finally
-          (fun () ->
-            Vhd_qcow_parsing.run_tool tool progress_cb args
-              ~output_fd:pipe_writer ~replace_fds
-          )
-          (fun () -> Unix.close pipe_writer)
-      )
-      ()
-  in
-
-  let map_pipe_reader, map_pipe_writer = Unix.pipe ~cloexec:true () in
-  let (_ : Thread.t) =
-    run_in_thread !Xapi_globs.qemu_img
-      ["map"; qcow_path; "--output=json"]
-      map_pipe_writer []
-  in
-
-  let info_pipe_reader, info_pipe_writer = Unix.pipe ~cloexec:true () in
-  let (_ : Thread.t) =
-    run_in_thread !Xapi_globs.qemu_img
-      ["info"; qcow_path; "--output=json"]
-      info_pipe_writer []
-  in
-
-  (map_pipe_reader, info_pipe_reader)
-
-let parse_header qcow_path =
-  let pipe, _ = read_header qcow_path in
-  Vhd_qcow_parsing.parse_header pipe
-
-let parse_header_interval qcow_path =
-  let pipes = read_header qcow_path in
-  Vhd_qcow_parsing.parse_header_qemu_img pipes
-
 let send ?relative_to (progress_cb : int -> unit) (unix_fd : Unix.file_descr)
     (path : string) (_size : Int64.t) =
   let qcow_of_device =
@@ -69,7 +30,9 @@ let send ?relative_to (progress_cb : int -> unit) (unix_fd : Unix.file_descr)
 
   (* If VDI is backed by QCOW, parse the header to determine nonzero clusters
      to avoid reading all of the raw disk *)
-  let input_fds = Result.map read_header qcow_path |> Result.to_option in
+  let input_fds =
+    Result.map Vhd_qcow_parsing.Qcow.read_header qcow_path |> Result.to_option
+  in
 
   (* TODO: If VHD headers are to be consulted as well, qcow2-to-stdout
      needs to properly account for cluster_bits. Currently QCOW2 export
@@ -84,7 +47,9 @@ let send ?relative_to (progress_cb : int -> unit) (unix_fd : Unix.file_descr)
     | None ->
         None
   in
-  let diff_fds = Option.map read_header relative_to_qcow_path in
+  let diff_fds =
+    Option.map Vhd_qcow_parsing.Qcow.read_header relative_to_qcow_path
+  in
 
   let map_fd_string = Uuidx.(to_string (make ())) in
   let info_fd_string = Uuidx.(to_string (make ())) in
diff --git a/ocaml/xapi/qcow_tool_wrapper.mli b/ocaml/xapi/qcow_tool_wrapper.mli
index 16cede3bb..51c3c6265 100644
--- a/ocaml/xapi/qcow_tool_wrapper.mli
+++ b/ocaml/xapi/qcow_tool_wrapper.mli
@@ -23,7 +23,3 @@ val send :
   -> string
   -> int64
   -> unit
-
-val parse_header : string -> int * int list
-
-val parse_header_interval : string -> int * (int * int) list
diff --git a/ocaml/xapi/stream_vdi.ml b/ocaml/xapi/stream_vdi.ml
index 4481cb767..641aeecf7 100644
--- a/ocaml/xapi/stream_vdi.ml
+++ b/ocaml/xapi/stream_vdi.ml
@@ -315,7 +315,7 @@ let send_one ofd (__context : Context.t) rpc session_id progress refresh_session
                   | "vhd" ->
                       Vhd_tool_wrapper.parse_header path
                   | "qcow2" ->
-                      Qcow_tool_wrapper.parse_header path
+                      Vhd_qcow_parsing.Qcow.parse_header path
                   | _ ->
                       failwith (Printf.sprintf "%s: unreachable" __FUNCTION__)
                 in
@@ -344,7 +344,7 @@ let send_one ofd (__context : Context.t) rpc session_id progress refresh_session
                   | "vhd" ->
                       Vhd_tool_wrapper.parse_header_interval path
                   | "qcow2" ->
-                      Qcow_tool_wrapper.parse_header_interval path
+                      Vhd_qcow_parsing.Qcow.parse_header_interval path
                   | _ ->
                       failwith (Printf.sprintf "%s: unreachable" __FUNCTION__)
                 in
diff --git a/opam/vhd-tool.opam b/opam/vhd-tool.opam
index 0edc71774..d1f89de71 100644
--- a/opam/vhd-tool.opam
+++ b/opam/vhd-tool.opam
@@ -35,6 +35,7 @@ depends: [
   "uri"
   "vhd-format" {= version}
   "vhd-format-lwt" {= version}
+  "vhd_qcow_parsing" {= version}
   "xapi-idl" {= version}
   "xapi-log" {= version}
   "xen-api-client-lwt" {= version}
diff --git a/opam/vhd_qcow_parsing.opam b/opam/vhd_qcow_parsing.opam
new file mode 100644
index 000000000..068192870
--- /dev/null
+++ b/opam/vhd_qcow_parsing.opam
@@ -0,0 +1,36 @@
+# This file is generated by dune, edit dune-project instead
+opam-version: "2.0"
+synopsis: """
+Wrapper to parse VHD and QCOW headers to determine allocated
+           clusters"""
+maintainer: ["Xapi project maintainers"]
+authors: ["xen-api@lists.xen.org"]
+license: "LGPL-2.1-only WITH OCaml-LGPL-linking-exception"
+homepage: "https://xapi-project.github.io/"
+bug-reports: "https://github.com/xapi-project/xen-api/issues"
+depends: [
+  "dune" {>= "3.20"}
+  "ocaml" {>= "4.13.0"}
+  "forkexec" {= version}
+  "xapi-consts" {= version}
+  "xapi-log" {= version}
+  "xapi-stdext-pervasives" {= version}
+  "yojson"
+  "odoc" {with-doc}
+]
+build: [
+  ["dune" "subst"] {dev}
+  [
+    "dune"
+    "build"
+    "-p"
+    name
+    "-j"
+    jobs
+    "@install"
+    "@runtest" {with-test}
+    "@doc" {with-doc}
+  ]
+]
+dev-repo: "git+https://github.com/xapi-project/xen-api.git"
+x-maintenance-intent: ["(latest)"]
diff --git a/opam/xapi.opam b/opam/xapi.opam
index 34dd9af3b..9af7ce8b4 100644
--- a/opam/xapi.opam
+++ b/opam/xapi.opam
@@ -93,6 +93,7 @@ depends: [
   "xenstore_transport"
   "xmlm"
   "xml-light2" {= version}
+  "vhd_qcow_parsing" {= version}
   "yojson"
   "zstd" {= version}
   "odoc" {with-doc}
